feat: add a way to limit the number of subscriptions per connection

niklasad1 opened the pull request:
Closing #729

niklasad1 committed:
feat: limit the number of subscriptions

Closing #729

niklasad1 committed:
fix nit

niklasad1 committed:
Update core/src/server/helpers.rs

jsdw commented on core/src/server/helpers.rs:
```diff
@@ -196,8 +198,41 @@ pub async fn collect_batch_response(rx: mpsc::UnboundedReceiver<String>) -> Stri
+/// Wrapper over [`tokio::sync::Notify`] with bounds check.
+#[derive(Debug)]
+pub struct BoundedSubscriptions {
+	inner: Arc<Notify>,
```
I'm wondering whether we need the `Arc` here? What would happen if we ditched it?

jsdw commented on core/src/server/helpers.rs:
```diff
@@ -196,8 +198,41 @@ pub async fn collect_batch_response(rx: mpsc::UnboundedReceiver<String>) -> Stri
 
+/// Wrapper over [`tokio::sync::Notify`] with bounds check.
+#[derive(Debug)]
+pub struct BoundedSubscriptions {
```
I was just sortof wodnering whether this could be replaced with a Semaphore (https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html), which would start at `max_subscriptions` and hand out permits for subscriptions until it runs out (and starts returning errors).

niklasad1 replied:
yeah, that would be nice the hard to part is where to store the semaphore after acquiring it.

I guess we could move it to the callbacks but that's quite annoying.

jsdw replied:
I'm happy to not worry about it for now; we can always have a pass later to see whether we can simplify/optimise things!

jsdw reviewed (approved):
Looks good to me!

Would be good to have an integration test that it works :)

niklasad1 committed:
add integration tests + some fixes so it works

niklasad1 committed:
Merge remote-tracking branch 'origin/master' into na-fix-729

niklasad1 committed:
cargo fmt

niklasad1 committed:
fix doc links

jsdw committed:
Unsubscribe calls should avoid subscription limits

Point to Tokio 1.16 (we use a method from it), and a little special treatment for unsubscribe methods

jsdw committed:
No resource limiting for Unsubscribe calls

jsdw committed:
Test that we can still unsubscribe after hitting a limit

lexnv commented on core/src/server/rpc_module.rs:
```diff
@@ -70,15 +72,15 @@ pub type ConnectionId = usize;
 	pub conn_id: ConnectionId,
-	/// Get notified when the connection to subscribers is closed.
-	pub close_notify: Arc<Notify>,
+	/// Get notified when the connection to subscribers is closed.1
```
```suggestion
	/// Get notified when the connection to subscribers is closed.
```

jsdw committed:
Fix a comment typo

Co-authored-by: Alexandru Vasile <60601340+lexnv@users.noreply.github.com>

niklasad1 commented on core/src/server/rpc_module.rs:
```diff
@@ -393,17 +405,19 @@ impl Methods {
-		let notify = Arc::new(Notify::new());
+		let bounded_subs = BoundedSubscriptions::new(u32::MAX);
+		let close_notify = bounded_subs.acquire().unwrap();
+		let notify = bounded_subs.acquire().unwrap();
```
```suggestion
		let notify = bounded_subs.acquire().expect("u32::MAX permits is sufficient; qed");
```

niklasad1 committed:
Update core/src/server/rpc_module.rs

niklasad1 committed:
Update core/src/server/rpc_module.rs
